### DISCUSSION TITLE
Each page has db connection VS Only api route has db connection

### DISCUSSION THREAD

**Discussion by bwesner** [vercel/next.js#18208] [Help] [answered]
I have an api route `/api` that the client will hit when it wants data.

I have other pages: `/page1`, `/page2` that want to load data from the db in their getInitialProps. Should I have `/page1` and `/page2` connect to the db to get their data when they are rendering on the server? Or should only `/api` connect to the db and `/page1` and `/page2` should simply send an http request to `/api` to get the data?

Note that I am using apollo, so `SchemaLink` (https://www.apollographql.com/docs/link/links/schema/) makes the request handling on the part of any page easy (i.e. i could use the same graphql schema for `/api`, `/page1`, and `/page2`). I'm just not sure if opening a lot more connections to the db (1 for every page that needs db data) is actually considered better than making the http request from each page to the `/api` route.

**Comment by juzhiyuan** [comment]
both ways are ok for me, but I would prefer fetching data from `/api` no matter in `getInitialProps (SSR)` or in react.

**Comment by jamesmosier** [comment, marked as answer]
You should make the database query directly in `getInitialProps`. Then you don't have to rely on the round trip time to call the API and return the response. I'd also recommend looking at `getServerSideProps` or `getStaticProps` instead if you are able to use those! It is also recommended that you write all your server logic directly as opposed to calling your API routes.

From the [docs](https://nextjs.org/docs/basic-features/data-fetching):

> You should not use `fetch()` to call an API route in your application. Instead, directly import the API route and call its function yourself. You may need to slightly refactor your code for this approach.

**Reply by bwesner** [reply]
Makes sense to me. My only concern is that it means more connections open to the db simultaneously but I guess that will generally be less of a bottleneck than having do to the roundtrip for the api call
